datafusion/common_runtime/src/common.rs

```diff
@@ -51,10 +51,22 @@ impl<R: 'static> SpawnedTask<R> {
         Self { inner }
     }
 
+    /// Joins the task, returning the result of join (`Result<R, JoinError>`).
     pub async fn join(mut self) -> Result<R, JoinError> {
         self.inner
             .join_next()
             .await
             .expect("`SpawnedTask` instance always contains exactly 1 task")
     }
+
+    /// Joins the task and unwinds the panic if it happens.
+    pub async fn join_unwind(self) -> R {
+        self.join().await.unwrap_or_else(|e| {
+            if e.is_panic() {
+                std::panic::resume_unwind(e.into_panic());
+            } else {
+                unreachable!();
```

Comment:
whi is this unreachable? Is there no other way for a Task::join to error?

Reply:
Hm, that's the good question. I just kept the logic as it used to be.

But in fact `JoinError` might be caused either by panic or cancellation of task.

And I think it's reachable in general, but if it was cancelled - likely it's expected or was due to another propagated error. 

For now I can suggest to remove `else { unreachable!() }`, but in long term we can think about smarter error handling for such cases

Reply:
I've updated this block, because actually it's really confusing. Added some notes to code

Reply:
I've update added a comment to the code to describe the behavior.

But in general as @devinjdangelo mentioned, we could improve error handling in the future